modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,616 @@
+    return inArg<GOAKContext::InputPtr*>(input);
+}
+
+GOAKContext::OutputPtr* GOAKContext::out(int output) {
+    return &(m_outputs.at(output));
+}
+
+namespace detail {
+template<class T> struct get_in;
+template<> struct get_in<cv::GFrame> {
+    static GOAKContext::InputPtr* get(GOAKContext &ctx, int idx) { return ctx.in(idx); }
+};
+template<class T> struct get_in {
+    static T get(GOAKContext &ctx, int idx) { return ctx.inArg<T>(idx); }
+};
+// FIXME: add support of other types
+
+template<class T> struct get_out;
+template<> struct get_out<cv::GFrame> {
+    static GOAKContext::OutputPtr* get(GOAKContext &ctx, int idx) { return ctx.out(idx); }
+};
+template<typename U> struct get_out<cv::GArray<U>> {
+    static GOAKContext::OutputPtr* get(GOAKContext &ctx, int idx) { return ctx.out(idx); }
+};
+// FIXME: add support of other types
+
+template<typename, typename, typename>
+struct OAKCallHelper;
+
+template<typename Impl, typename... Ins, typename... Outs>
+struct OAKCallHelper<Impl, std::tuple<Ins...>, std::tuple<Outs...> > {
+    template<int... IIs, int... OIs>
+    static void construct_impl(  GOAKContext &ctx
+                               , std::shared_ptr<dai::Node>& node
+                               , std::vector<std::pair<std::string, dai::Buffer>>& in_queues_params
+                               , cv::detail::Seq<IIs...>
+                               , cv::detail::Seq<OIs...>) {
+        Impl::put(ctx.pipeline(),
+                  ctx.camera_size(),
+                  node,
+                  in_queues_params,
+                  get_in<Ins>::get(ctx, IIs)...,
+                  get_out<Outs>::get(ctx, OIs)...);
+    }
+
+    static void construct(GOAKContext &ctx,
+                          std::shared_ptr<dai::Node>& node,
+                          std::vector<std::pair<std::string, dai::Buffer>>& in_queues_params) {
+        construct_impl(ctx,
+                       node,
+                       in_queues_params,
+                       typename cv::detail::MkSeq<sizeof...(Ins)>::type(),
+                       typename cv::detail::MkSeq<sizeof...(Outs)>::type());
+    }
+};
+
+} // namespace detail
+
+struct GOAKKernel {
+    using F = std::function<void(GOAKContext&,
+                                 std::shared_ptr<dai::Node>&,
+                                 std::vector<std::pair<std::string, dai::Buffer>>&)>;
+    explicit GOAKKernel(const F& f) : m_f(f) {}
+    const F m_f;
+};
+
+struct OAKComponent
+{
+    static const char *name() { return "OAK Component"; }
+    GOAKKernel k;
+};
+
+}} // namespace gimpl // namespace cv
+
+using OAKGraph = ade::TypedGraph
+    < cv::gimpl::OAKComponent
+    // FIXME: extend
+    >;
+
+using ConstOAKGraph = ade::ConstTypedGraph
+    < cv::gimpl::OAKComponent
+    // FIXME: extend
+    >;
+
+void cv::gimpl::GOAKExecutable::LinkToParentHelper(ade::NodeHandle handle,
+                                                   const std::vector<ade::NodeHandle>& nodes)
```

Comment:
please document what this function does - at a high level

Reply:
Done